### ISSUE TITLE
balena push: Could not detect project type: Service main: Couldn't satisfy node version >=4 

### ISSUE THREAD

**Issue by jiantaolovebingbing** [balena-io/balena-cli#1819]
- **balena CLI version:** e.g. 11.2.1 (output of the `"balena version"` command)
- **Operating system version:** e.g. Windows 10, Ubuntu 18.04, macOS 10.14.5
- **32/64 bit OS and processor:** e.g. 32-bit Windows on 64-bit Intel processor
- **Install method:** npm or zip or executable installer
- **If npm install, Node.js and npm version:** e.g. Node v8.16.0 and npm v6.4.1

1. balena CLI version: 11.33.3
- Operating system version: macOS 10.15.4
- 32/64 bit OS and processor: MacBook Pro 16-inch, 2019, so 64 bit
- Install method: release
- Node: v10.16, npm: v6.9.0

![image](https://user-images.githubusercontent.com/33253363/81686743-42c82780-948b-11ea-99e5-b325320b0628.png)

**Comment by jiantaolovebingbing** [comment]
![image](https://user-images.githubusercontent.com/33253363/81690480-e6b2d280-948d-11ea-87f9-4268c84376ed.png)

**Comment by pdcastro** [comment]
@jiantaolovebingbing, thanks for reporting this issue.
A balena project (application) is most commonly defined with a `Dockerfile`, but may be also be defined with a `package.json` file, as documented at:
* https://www.balena.io/docs/learn/develop/dockerfile/
* https://www.balena.io/docs/learn/develop/dockerfile/#node-applications

The error message suggests that your project uses a `package.json` file instead of a `Dockerfile`. Is this right?

To try and reproduce the error you were getting, I created a folder with a trivial `package.json` file with these contents:
```
{
  "engines": {
    "node": ">=4.0"
  }
}
```
Then I tested the `balena push` command with that folder, against a Raspberry Pi 3 (32 bits), a Raspberry Pi 4 (64 bits) and an Intel NUC. I was able to reproduce the error with the Raspberry Pi 4 only:
```
$ ./bin/balena push 192.168.0.37 -s ~/test-project/
[Info]    Starting build on device 192.168.0.37
[Info]    No "docker-compose.yml" file found at "/Users/paulo/test-project/"
[Info]    Creating default composition with source: "/Users/paulo/test-project/"
Couldn't satisfy node version >=4
```
With a Raspberry Pi 3 or the Intel NUC, the error did not happen:
```
$ ./bin/balena push 192.168.0.21 -s ~/balena/other/simple_container/
[Info]    Starting build on device 192.168.0.21
[Info]    No "docker-compose.yml" file found at "/Users/paulo/balena/other/simple_container/"
[Info]    Creating default composition with source: "/Users/paulo/balena/other/simple_container/"
[Build]   [main] Step 1/4 : FROM resin/intel-nuc-node:10.9.0-onbuild
[Build]   [main] # Executing 3 build trigger
...
[Build]   [main] Step 2/4 : RUN ln -s /usr/src/app /app
[Build]   [main]  ---> Running in 829358eeb345
[Build]   [main] Removing intermediate container 829358eeb345
[Build]   [main]  ---> f76e4a5b42bd
[Build]   [main] Step 3/4 : LABEL io.resin.local.image=1
[Build]   [main]  ---> Running in 529abf43623a
[Build]   [main] Removing intermediate container 529abf43623a
[Build]   [main]  ---> 2ccbcefc8b25
[Build]   [main] Step 4/4 : LABEL io.resin.local.service=main
[Build]   [main]  ---> Running in e580393619d4
[Build]   [main] Removing intermediate container e580393619d4
[Build]   [main]  ---> a11dba059d11
[Build]   [main] Successfully built a11dba059d11
[Build]   [main] Successfully tagged local_image_main:latest

[Live]    Waiting for device state to settle...
[Info]    Streaming device logs...
[Live]    Watching for file changes...
...
[Logs]    [5/12/2020, 16:52:04] Starting service 'main sha256:a11dba059d11f3d179f5dd1c22267fe18029ebcd5b01652c17de4528e2a35c73'
```

Which device type are you using? (Like Raspberry Pi 4, or something else.)

Overall, also for performance reasons, if your project does not have a `Dockerfile`, I would recommend that you created one, following the documentation links above. But I will wait for your reply before writing any further. :-)

**Comment by jiantaolovebingbing** [comment]
My device type is Raspberry Pi 3 model B+. But I don't add device, i think it would't affect the push action.
![image](https://user-images.githubusercontent.com/33253363/81764813-ffaa9a80-9504-11ea-9930-7302f928919f.png)

**Comment by jiantaolovebingbing** [comment]
![IMG_5655](https://user-images.githubusercontent.com/33253363/81764958-45fff980-9505-11ea-81a4-8cbae9506039.JPG)

**Comment by jiantaolovebingbing** [comment]
I have added the device, but still the same error. One thing i noticed, my device is 32bit, my OS file is based  on Raspberry Pi 3(using 64bit os), does it affect?
![image](https://user-images.githubusercontent.com/33253363/81767404-ea386f00-950a-11ea-9022-17a6eb0d4ce2.png)

**Comment by jiantaolovebingbing** [comment]
Yes, my project uses a package.json file. I tried electron-based project and nodejs-based project,
all go wrong.

**Comment by pdcastro** [comment]
@jiantaolovebingbing, thank you for the additional information. The "workaround" suggestion for now is that you add a `Dockerfile.template` file to your project, for example [this file](https://github.com/balena-io-examples/balena-node-hello-world/blob/master/Dockerfile.template) from the [balena-node-hello-world](https://github.com/balena-io-examples/balena-node-hello-world/) sample project. More documentation about Dockerfiles in balena projects can be found at: https://www.balena.io/docs/learn/develop/dockerfile/

> my OS file is based on Raspberry Pi 3(using 64bit os), does it affect?

I can reproduce the error and I can confirm that it is a bug that needs to be fixed -- thank you again for reporting it. Meanwhile, please add a `Dockerfile.template` file as suggested above. It should then work regardless of the device type.

**Comment by jiantaolovebingbing** [comment]
thanks a lot! It works.But there are other problem during the push.
I installed the python environment, but I get an error python not found in path. Should I install python 3?
![image](https://user-images.githubusercontent.com/33253363/81810945-6efcaa80-9556-11ea-888a-34e8af108be1.png)

**Comment by jiantaolovebingbing** [comment]
![image](https://user-images.githubusercontent.com/33253363/81810991-7e7bf380-9556-11ea-8b5c-4cbed882f323.png)

**Comment by pdcastro** [comment]
> I get an error python not found in path

I believe this is now an issue with getting the Snapmakerjs project to work with a balenalib base image, rather than a balena CLI bug. If you need help with setting up the Snapmakerjs project, I suggest creating a thread in the balena support forums, which is monitored by balena's support team and the wider community: https://forums.balena.io/c/balena-cloud/
Thanks again.

**Comment by jiantaolovebingbing** [comment]
Never mind! my fault, just changing the `Dockerfile.template` file can fix it! thanks again.
But how could the status is still inactive?
![image](https://user-images.githubusercontent.com/33253363/81815193-47104580-955c-11ea-8535-2b14a6147fac.png)
![image](https://user-images.githubusercontent.com/33253363/81815283-60b18d00-955c-11ea-9e00-061d97d45a1c.png)

**Comment by pdcastro** [comment]
> But how could the status is still inactive?

It looks like the device did not boot successfully, or perhaps a local network firewall is blocking the VPN link. See [Network Requirements](https://www.balena.io/docs/reference/OS/network/1.x/#network-requirements). For further troubleshooting help, I suggest creating a thread in the balena support forums, which is monitored by balena's support team and the wider community: https://forums.balena.io/c/balena-cloud/
Thanks again.

**Comment by ccimpoi** [comment]
I had the same issue stated in the issue title and my problem was that by mistake I was running balena push from the app folder where I had the app package.json. I should ve run the command from the parent folder that held the Dockerfile.template. So Balena was trying to make sense of the package.json as a Dockerfile replacement which led to the correct but misleading error.
